cpp/test/sg/rf_depth_test.cu

```diff
@@ -0,0 +1,301 @@
+        data_h[row + col * params.n_rows] = d(gen);
+      }
+    }
+    updateDevice(data, data_h.data(), data_len, stream);
+
+    // Populate labels
+    labels_h.resize(params.n_rows);
+    for (int row = 0; row < params.n_rows; ++row) {
+      labels_h[row] =
+        (data_h[row + 2 * params.n_rows] * data_h[row + 3 * params.n_rows]);
+    }
+    updateDevice(labels, labels_h.data(), params.n_rows, stream);
+
+    forest = new typename ML::RandomForestMetaData<T, T>;
+    null_trees_ptr(forest);
+
+    cumlHandle handle(rf_params.n_streams);
+    handle.setStream(stream);
+
+    fit(handle, forest, data, params.n_rows, params.n_cols, labels, rf_params);
+
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+  }
+
+  void SetUp() override { basicTest(); }
+
+  void TearDown() override {
+    labels_h.clear();
+
+    CUDA_CHECK(cudaFree(labels));
+    CUDA_CHECK(cudaFree(data));
+    delete forest;
+  }
+
+ protected:
+  RfInputs<T> params;
+  T* data;
+  T* labels;
+  std::vector<T> labels_h;
+
+  RandomForestMetaData<T, T>* forest;
+};
+
+template <typename L, typename T>
+int MaxDepthOfDecisionTree(const DecisionTree::TreeMetaDataNode<T, L>* tree) {
+  const auto& node_array = tree->sparsetree;
+  std::queue<std::pair<int, int>> q;  // (node ID, depth)
+  // Traverse the tree breadth-first
+  int initial_depth = 1;
+  // Currently, cuML starts counting depth by 1.
+  // See https://github.com/rapidsai/cuml/issues/2518#issuecomment-660070910.
+  q.emplace(0, initial_depth);
+  int max_depth = initial_depth;
+  while (!q.empty()) {
+    int node_id, depth;
+    std::tie(node_id, depth) = q.front();
+    q.pop();
+    max_depth = std::max(depth, max_depth);
+    const SparseTreeNode<T, L>& node = node_array.at(node_id);
+    if (node.colid != -1) {
+      q.emplace(node.left_child_id, depth + 1);
+      q.emplace(node.left_child_id + 1, depth + 1);
+    }
+  }
+  return max_depth;
+}
+
+typedef RfClassifierDepthTest<float> RfClassifierDepthTestF;
+TEST_P(RfClassifierDepthTestF, Fit) {
+  CUML_LOG_INFO("Param max_depth = %d", params.max_depth);
+  for (int i = 0; i < forest->rf_params.n_trees; i++) {
+    int actual_max_depth = MaxDepthOfDecisionTree(&(forest->trees[i]));
+    ASSERT_EQ(actual_max_depth, params.max_depth);
+    ASSERT_EQ(actual_max_depth, forest->trees[i].depth_counter);
+  }
+}
+
+typedef RfClassifierDepthTest<double> RfClassifierDepthTestD;
+TEST_P(RfClassifierDepthTestD, Fit) {
+  CUML_LOG_INFO("Param max_depth = %d", params.max_depth);
+  for (int i = 0; i < forest->rf_params.n_trees; i++) {
+    int actual_max_depth = MaxDepthOfDecisionTree(&(forest->trees[i]));
+    ASSERT_EQ(actual_max_depth, params.max_depth);
+    ASSERT_EQ(actual_max_depth, forest->trees[i].depth_counter);
+  }
+}
+
+INSTANTIATE_TEST_CASE_P(RfClassifierDepthTests, RfClassifierDepthTestF,
+                        ::testing::Range(1, 20));
+
+INSTANTIATE_TEST_CASE_P(RfClassifierDepthTests, RfClassifierDepthTestD,
```

Comment:
Is it necessary to do every version up to 20? Does this get slow or still pretty instantaneous? If the test is too long, could consider doing like depths 1, 4, 7, and skipping up to 20 ish etc...

Reply:
It's quite fast. In the gpuCI, it takes only 2 seconds to run the whole suite. And it is important to test all cases from 1 to 20, to check for a possible off-by-one error. This time, the error was only applicable for max_depth > 14.

Reply:
Great, I think we can afford 2 seconds!